core/src/main/scala/org/apache/spark/shuffle/BlockStoreShuffleReader.scala

```diff
@@ -104,9 +104,18 @@ private[spark] class BlockStoreShuffleReader[K, C](
         context.taskMetrics().incMemoryBytesSpilled(sorter.memoryBytesSpilled)
         context.taskMetrics().incDiskBytesSpilled(sorter.diskBytesSpilled)
         context.taskMetrics().incPeakExecutionMemory(sorter.peakMemoryUsedBytes)
+        // Use completion callback to stop sorter if task was cancelled.
+        context.addTaskCompletionListener(tc => {
+          // Note: we only stop sorter if cancelled as sorter.stop wouldn't be called in
+          // CompletionIterator. Another way would be making sorter.stop idempotent.
+          if (tc.isInterrupted()) { sorter.stop() }
```

Comment:
seems we can remove this `if` if we don't return a `CompletionIterator`.

BTW I think we need to check all the places that use `CompletionIterator`, to see if they consider job canceling.

Reply:
One advantage of `CompletionIterator` is that the `completionFunction` will be called as soon as the wrapped iterator is consumed. So for sorter, it will release memory earlier rather than at task completion.

As for job cancelling, It's not just `CompletionIterator` that we should consider. The combiner and sorter pattern(or similar) is something we should look for:
``` scala
combiner.insertAll(iterator) // or sorter.insertAll(iterator)
// then returns new iterator
combiner.iterator // or sorter.iterator
```

Reply:
I may be missing something obvious, but seems `ExternalSorter.stop()` is already idempotent?

Reply:
> I may be missing something obvious, but seems ExternalSorter.stop() is already idempotent?

Ah, yes. After another look, it's indeed idempotent. 
Will update the code.